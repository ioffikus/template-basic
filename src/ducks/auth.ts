import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { push } from 'connected-react-router';
import { GET, POST } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'AUTH';
const NS = `${REDUCER}__`;

export interface IAuthState extends Store.IReduxCommonState {
  isAuthenticated: boolean;
  isAuthLinkRequesting: boolean;
  url: string;
  data: object;
}

export const initialState: IAuthState = {
  isAuthenticated: false,
  isRequesting: false,
  isRequestError: false,
  isAuthLinkRequesting: false,
  url: null,
  data: null,
};

const reducer = createReducer<IAuthState>({}, initialState);

export const loginUserSuccess = createAction(`${NS}LOGIN_USER_SUCCESS`);
reducer.on(loginUserSuccess, state => ({
  ...state,
  isRequesting: false,
  isRequestError: false,
  isAuthenticated: true,
}));

const authLinkSuccess = createAction<{ url: string; data: object }>(`${NS}AUTH_LINK_SUCCESS`);
reducer.on(authLinkSuccess, (state, { url, data }) => ({
  ...state,
  url,
  data: { ...data },
  isAuthLinkRequesting: false,
}));

const loginUserFailure = createAction(`${NS}LOGIN_USER_FAILURE`);
reducer.on(loginUserFailure, state => ({
  ...state,
  isRequesting: false,
  isAuthenticated: false,
  isRequestError: true,
}));

const authLinkFailure = createAction(`${NS}AUTH_LINK_FAILURE`);
reducer.on(authLinkFailure, state => ({
  ...state,
  isAuthLinkRequesting: false,
  url: null,
  data: null,
}));

const loginUserRequest = createAction(`${NS}LOGIN_USER_REQUEST`);
reducer.on(loginUserRequest, state => ({
  ...state,
  isRequesting: true,
}));

const authLinkRequest = createAction(`${NS}AUTH_LINK_REQUEST`);
reducer.on(authLinkRequest, state => ({
  ...state,
  isAuthLinkRequesting: true,
  url: null,
  data: null,
}));

const reset = createAction(`${NS}RESET`);
reducer.on(reset, () => ({
  ...initialState,
}));

export const getAuthLink = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(authLinkRequest());
    GET(
      {
        url: `${EndpointsConfig.AUTH.AUTH_LINK}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { url, data } = resp.data as { url: string; data: object };
        // WIP does not use Alicanto.API.GET_AUTH_OAUTH_PROVIDER_URL_RESPONSE for develop
        dispatch(authLinkSuccess({ url, data }));
        resolve();
      })
      .catch(error => {
        dispatch(authLinkFailure());
        reject(error);
      });
  });
};

export const loginUser = (username: string, password: string): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    dispatch(loginUserRequest());
    const data: Alicanto.API.POST_AUTH_LOGIN_BODY = {
      username,
      password,
    };
    POST(
      {
        data,
        url: `${EndpointsConfig.AUTH.LOGIN}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(loginUserSuccess());
        if (process.env.BROWSER) {
          // TODO: move to external
          // redirect to home page
          dispatch(push('/'));
        }
        resolve();
      })
      .catch(error => {
        dispatch(loginUserFailure());
        reject(error);
      });
  });
};

export const logoutUser = (cb: () => void): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(loginUserRequest());
    GET(
      {
        url: `${EndpointsConfig.AUTH.LOGOUT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(loginUserSuccess());
        dispatch(reset());
        if (cb) {
          cb();
        }
        if (process.env.BROWSER) {
          // TODO: move to external
          // redirect to home page
          dispatch(push('/'));
        }
        resolve();
      })
      .catch(error => {
        dispatch(loginUserFailure());
        reject(error);
      });
  });
};

export default reducer;
