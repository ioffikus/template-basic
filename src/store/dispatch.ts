import { Store } from 'redux';
import _ from 'lodash';
import { showError } from 'src/store/common';

let store: Store<Store.IState>;

export const setupDispatch = (config: { store: Store<Store.IState> }) => {
  store = config.store;
};

const dispatch = (action: any) => {
  if (_.isFunction(action)) {
    return store.dispatch(action).catch((error: any) => {
      if (!_.isNull(error)) {
        showError(error);
      }
      return Promise.reject(error);
    });
  }
  return store.dispatch(action);
};

export default dispatch;
