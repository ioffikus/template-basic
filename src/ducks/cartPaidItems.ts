import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { ICardPaidItemsState } from 'src/ducks/cartPaidItems';
import _ from 'lodash';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CART_PAID_ITEMS';
const NS = `${REDUCER}__`;

export interface ICardPaidItemsState extends Alicanto.API.GET_CART_PAID_ITEMS_RESPONSE, Store.IReduxCommonState {}

export const initialState: ICardPaidItemsState = {
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequestError: false,
  isRequesting: false,
};

const reducer = createReducer<ICardPaidItemsState>({}, initialState);

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequestError: true,
  isRequesting: false,
}));

const callSuccess = createAction<Alicanto.API.GET_CART_PAID_ITEMS_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { next, previous, count, results }) => ({
  ...state,
  next,
  previous,
  count,
  results: [...results],
  isRequestError: false,
  isRequesting: false,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

interface ICartPaidItemsQuery {
  page?: string;
  page_size?: string;
}

const getCartPaidItemsRequestParams = ({ page, page_size }: ICartPaidItemsQuery) => {
  const params: ICartPaidItemsQuery = {
    page_size: '10',
  };
  if (!_.isNil(page)) {
    params.page = page;
  }
  if (!_.isNil(page_size)) {
    params.page_size = page_size;
  }
  return params;
};

export const getCartPaidItems = (page = '1'): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_PAID_ITEMS}`,
        params: getCartPaidItemsRequestParams({ page }),
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CART_PAID_ITEMS_RESPONSE;
        let results = data.results;
        if (page !== '1') {
          const oldResults = getState().cartPaidItems.results;
          results = [...oldResults, ...results];
        }
        dispatch(callSuccess({ ...data, results }));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
