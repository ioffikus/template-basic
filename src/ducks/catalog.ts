import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import _ from 'lodash';
import { ICatalogQueryState } from 'src/ducks/catalogQuery';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CATALOG';
const NS = `${REDUCER}__`;

export interface ICatalogState extends Alicanto.API.GET_CATALOG_RESPONSE, Store.IReduxCommonState {}

export const initialState: ICatalogState = {
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<ICatalogState>({}, initialState);

const callSuccess = createAction<Alicanto.API.GET_CATALOG_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { next, previous, count, results }) => ({
  ...state,
  next,
  previous,
  count,
  results: [...results],
  isRequesting: false,
  isRequestError: false,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

interface ICatalogQuery {
  search?: string;
  page?: string;
  page_size: string;
  ordering?: string;
  filter?: string;
}

const getCatalogRequestParams = ({ search, page, pageSize, ordering, filterObj }: ICatalogQueryState) => {
  const params: ICatalogQuery = {
    page_size: pageSize,
  };
  if (!_.isEmpty(search)) {
    params.search = search;
  }
  if (!_.isEmpty(page)) {
    params.page = page;
  }
  if (!_.isEmpty(ordering)) {
    params.ordering = ordering;
  }
  if (!_.isEmpty(filterObj)) {
    params.filter = JSON.stringify(filterObj);
  }
  return params;
};

export const getCatalog = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    const catalogQuery = getState().catalogQuery;
    GET(
      {
        url: EndpointsConfig.CATALOG,
        params: getCatalogRequestParams(catalogQuery),
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_RESPONSE;
        let results = data.results;
        if (catalogQuery.page !== '1') {
          const oldResults = getState().catalog.results;
          results = [...oldResults, ...results];
        }
        dispatch(callSuccess({ ...data, results }));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
