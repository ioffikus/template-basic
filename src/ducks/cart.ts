import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import _ from 'lodash';
import { BusinessLogicError } from 'src/store/common';
import { safelyParseJSON } from 'src/helpers/safelyParseJSON';
import { IObject } from 'src/core/interfaces/IObject';
import { GET, POST, PUT, DELETE } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CART';
const NS = `${REDUCER}__`;

export interface ICartState extends Store.IReduxCommonState {
  id: number;
  items: Alicanto.Models.CatalogItem[];
  itemsCount: number;
  totalPrice: number;
}

export const initialState: ICartState = {
  id: null,
  items: [],
  itemsCount: 0,
  totalPrice: 0,
  isRequestError: false,
  isRequesting: false,
};

const reducer = createReducer<ICartState>({}, initialState);

const addProductToCart = createAction<Alicanto.Models.CatalogItem>(`${NS}ADD_PRODUCT_TO_CART`);
reducer.on(addProductToCart, (state, product) => ({
  ...state,
  items: [
    ...state.items,
    {
      ...product,
    },
  ],
  itemsCount: state.itemsCount + 1,
  totalPrice: (state.totalPrice * 100 + product.schema_fields.price.value * 100) / 100,
}));

const deleteProductFromCart = createAction<number>(`${NS}DELETE_PRODUCT_FROM_CART`);
reducer.on(deleteProductFromCart, (state, index) => ({
  ...state,
  items: state.items.filter((_, i) => i !== index),
  itemsCount: state.itemsCount - state.items[index].count,
  totalPrice: (state.totalPrice * 100 - state.items[index].schema_fields.price.value * 100) / 100,
}));

const increaseProductQuantity = createAction<number>(`${NS}INCREASE_PRODUCT_QUANTITY`);
reducer.on(increaseProductQuantity, (state, index) => ({
  ...state,
  items: state.items
    .slice(0, index)
    .concat([
      {
        ...state.items[index],
        count: state.items[index].count + 1,
      },
    ])
    .concat(state.items.slice(index + 1)),
  itemsCount: state.itemsCount + 1,
  totalPrice: (state.totalPrice * 100 + state.items[index].schema_fields.price.value * 100) / 100,
}));

const decreaseProductQuantity = createAction<number>(`${NS}DECREASE_PRODUCT_QUANTITY`);
reducer.on(decreaseProductQuantity, (state, index) => ({
  ...state,
  items: state.items
    .slice(0, index)
    .concat([
      {
        ...state.items[index],
        count: state.items[index].count - 1,
      },
    ])
    .concat(state.items.slice(index + 1)),
  itemsCount: state.itemsCount - 1,
  totalPrice: (state.totalPrice * 100 - state.items[index].schema_fields.price.value * 100) / 100,
}));

const blockItemIncrease = createAction<number>(`${NS}BLOCK_ITEM_INCREASE`);
reducer.on(blockItemIncrease, (state, index) => ({
  ...state,
  items: state.items
    .slice(0, index)
    .concat([
      {
        ...state.items[index],
        blockedIncreasing: true,
      },
    ])
    .concat(state.items.slice(index + 1)),
}));

const putProductsInStore = createAction<Alicanto.Models.CatalogItem[]>(`${NS}PUT_PRODUCTS_IN_STORE`);
reducer.on(putProductsInStore, (state, items) => ({
  ...state,
  items: [...items],
}));

const putCartInStore = createAction<{ id: number; itemsCount: number; totalPrice: number }>(`${NS}PUT_CART_IN_STORE`);
reducer.on(putCartInStore, (state, { id, itemsCount, totalPrice }) => ({
  ...state,
  id,
  itemsCount,
  totalPrice,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  isRequestError: true,
  isRequesting: false,
}));

const callSuccess = createAction(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, state => ({
  ...state,
  isRequestError: false,
  isRequesting: false,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequestError: false,
  isRequesting: true,
}));

export const getCartFromLocalStorage = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  if (process.env.BROWSER) {
    const s = window.localStorage.getItem('cart');
    if (s) {
      const localStorageCartState: ICartState = safelyParseJSON(s);
      if (!_.isNull(localStorageCartState)) {
        const { id, itemsCount, totalPrice, items } = localStorageCartState;
        dispatch(putCartInStore({ id, itemsCount, totalPrice }));
        dispatch(putProductsInStore(items));
      }
    }
  }
  return Promise.resolve();
};

const setCartInLocalStorage = (cartState: ICartState) => {
  if (process.env.BROWSER) {
    const { items, itemsCount, totalPrice, id } = cartState;
    window.localStorage.setItem('cart', JSON.stringify({ items, itemsCount, totalPrice, id }));
  }
};

const removeCartFromLocalStorage = () => {
  if (process.env.BROWSER) {
    window.localStorage.removeItem('cart');
  }
};

const getProducts = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_CURRENT_ITEMS}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { results } = resp.data as Alicanto.API.GET_CART_CURRENT_ITEMS_RESPONSE;
        dispatch(putProductsInStore(results));
        dispatch(callSuccess());
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

const getCart = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_CURRENT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { items_count, total, id } = resp.data as Alicanto.API.GET_CART_CURRENT_RESPONSE;
        dispatch(putCartInStore({ id, itemsCount: items_count, totalPrice: parseFloat(total) }));
        dispatch(callSuccess());
        if (!items_count) {
          dispatch(deleteCart());
          resolve();
          return;
        }
        return dispatch(getProducts()).then(() => {
          setCartInLocalStorage(getState().cart);
          resolve();
        });
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

/**
 * If unauthenticated user've added some products, then
 * after login we put items on server,
 * firstly check if there are repeated items
 * in local storage and server. After that update cart info
 */
const putCart = (data: Alicanto.API.PUT_CART_CURRENT_BODY): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    PUT(
      {
        data,
        url: `${EndpointsConfig.CART_CURRENT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(callSuccess());
        return dispatch(getCart()).then(resolve);
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

const findUnicItems = (itemsInStore: Alicanto.Models.CatalogItem[], itemsInCart: Alicanto.Models.CatalogItem[]) => {
  const cartIds: IObject<Alicanto.Models.CatalogItem> = {};
  itemsInCart.forEach((item: Alicanto.Models.CatalogItem) => {
    cartIds[item.item] = item;
  });
  return itemsInStore.filter(item => !(item.item in cartIds));
};

export const postLocalStorageCart = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    if (_.isEmpty(items)) {
      dispatch(getCart()).then(resolve, reject);
      return;
    }
    dispatch(callRequest());
    GET(
      {
        url: `${EndpointsConfig.CART_CURRENT_ITEMS}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { results } = resp.data as Alicanto.API.GET_CART_CURRENT_ITEMS_RESPONSE;
        const itemsInCart = results;
        const unicItemsInStore = findUnicItems(items, itemsInCart);
        const data: Alicanto.API.PUT_CART_CURRENT_BODY = {
          items: [],
        };
        if (!_.isEmpty(unicItemsInStore)) {
          unicItemsInStore.forEach(item => {
            data.items.push({
              count: item.count,
              item: item.item,
            });
          });
          return dispatch(putCart(data)).then(resolve);
        }
        dispatch(callSuccess());
        return dispatch(getCart()).then(resolve);
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

/**
 * Dispatch action after click on 'add to card' button:
 * first click will add item in the cart
 * user can press the button multiple times and this action will increase product quantity
 * @param product
 */
export const addToCart = (product: Alicanto.Models.CatalogItem, cb: () => void): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    const productBalance = product.balance;
    const productIsUnlimited = product.is_unlimited;
    const productUid = product.uid;
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    const errorMsg = getState().ui.routes.cart.errors.moreThanInBalance.i18n.defaultText;
    const isAuthenticated = getState().auth.isAuthenticated;
    if (isAuthenticated) {
      dispatch(callRequest());
      if (itemIndex !== -1) {
        const item = items[itemIndex];
        if (productBalance < item.count + 1 && !productIsUnlimited) {
          dispatch(callFailure());
          const error = new BusinessLogicError(errorMsg);
          reject(error);
          return;
        }
        const data: Alicanto.API.PUT_CART_CURRENT_ITEMS_ID_BODY = {
          count: item.count + 1,
          item: productUid,
        };
        PUT(
          {
            data,
            url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
          },
          xShopHeaders,
        )
          .then((resp: AxiosResponse) => {
            dispatch(increaseProductQuantity(itemIndex));
            setCartInLocalStorage(getState().cart);
            dispatch(callSuccess());
            resolve();
          })
          .catch(error => {
            dispatch(callFailure());
            reject(error);
          });
        return;
      }
      const data: Alicanto.API.POST_CART_CURRENT_ITEMS_BODY = {
        count: 1,
        item: productUid,
      };
      POST(
        {
          data,
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          const data = resp.data as Alicanto.API.POST_CART_CURRENT_ITEMS_RESPONSE;
          dispatch(addProductToCart(data));
          if (cb) {
            cb();
          }
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    if (itemIndex !== -1) {
      dispatch(callRequest());
      GET(
        {
          url: `${EndpointsConfig.PRODUCT}${product.uid}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          const data = resp.data as Alicanto.API.GET_CATALOG_UID_RESPONSE;
          const item = items[itemIndex];
          if (item.count + 1 < data.balance || data.is_unlimited) {
            dispatch(increaseProductQuantity(itemIndex));
            if (cb) {
              cb();
            }
            setCartInLocalStorage(getState().cart);
            dispatch(callSuccess());
            resolve();
            return;
          }
          const error = new BusinessLogicError(errorMsg);
          return Promise.reject(error);
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    const newProduct = { ...product };
    newProduct.count = 1;
    newProduct.item = product.uid;
    delete newProduct.uid;
    dispatch(addProductToCart(newProduct));
    if (cb) {
      cb();
    }
    setCartInLocalStorage(getState().cart);
    resolve();
  });
};

export const deleteFromCart = (productUid: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    const item = items[itemIndex];
    const isAuthenticated = getState().auth.isAuthenticated;
    if (isAuthenticated) {
      dispatch(callRequest());
      DELETE(
        {
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          dispatch(deleteProductFromCart(itemIndex));
          if (getState().cart.itemsCount === 0) {
            dispatch(deleteCart());
          } else {
            setCartInLocalStorage(getState().cart);
          }
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    dispatch(deleteProductFromCart(itemIndex));
    if (getState().cart.itemsCount === 0) {
      dispatch(deleteCart());
    } else {
      setCartInLocalStorage(getState().cart);
    }
    resolve();
  });
};

export const increaseQuantity = (productUid: string, cb: () => void): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    if (itemIndex === -1) {
      resolve();
      return;
    }
    const item = items[itemIndex];
    const isAuthenticated = getState().auth.isAuthenticated;
    dispatch(callRequest());
    if (isAuthenticated) {
      const data: Alicanto.API.PUT_CART_CURRENT_ITEMS_ID_BODY = {
        count: item.count + 1,
        item: productUid,
      };
      PUT(
        {
          data,
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          dispatch(increaseProductQuantity(itemIndex));
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(blockItemIncrease(itemIndex));
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    GET(
      {
        url: `${EndpointsConfig.PRODUCT}${productUid}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_UID_RESPONSE;
        if (item.count + 1 < data.balance || data.is_unlimited) {
          dispatch(increaseProductQuantity(itemIndex));
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
          return;
        }
        const errorMsg = getState().ui.routes.cart.errors.moreThanInBalance.i18n.defaultText;
        return Promise.reject(new BusinessLogicError(errorMsg));
      })
      .catch(error => {
        dispatch(blockItemIncrease(itemIndex));
        dispatch(callFailure());
        reject(error);
      });
  })
    .then(() => {
      if (cb) {
        cb();
      }
    })
    .catch(error => {
      if (cb) {
        cb();
      }
      return Promise.reject(error);
    });
};

export const decreaseQuantity = (productUid: string, cb: () => void): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    const items = getState().cart.items;
    const itemIndex = items.findIndex(element => element.item === productUid);
    if (itemIndex === -1) {
      resolve();
      return;
    }
    const item = items[itemIndex];
    const isAuthenticated = getState().auth.isAuthenticated;
    if (isAuthenticated) {
      dispatch(callRequest());
      const data: Alicanto.API.PUT_CART_CURRENT_ITEMS_ID_BODY = {
        count: item.count - 1,
        item: productUid,
      };
      PUT(
        {
          data,
          url: `${EndpointsConfig.CART_CURRENT_ITEMS}${item.id}`,
        },
        xShopHeaders,
      )
        .then((resp: AxiosResponse) => {
          dispatch(decreaseProductQuantity(itemIndex));
          setCartInLocalStorage(getState().cart);
          dispatch(callSuccess());
          resolve();
        })
        .catch(error => {
          dispatch(callFailure());
          reject(error);
        });
      return;
    }
    // user isn't authenticated
    if (item.count > 1) {
      dispatch(decreaseProductQuantity(itemIndex));
      setCartInLocalStorage(getState().cart);
      resolve();
      return;
    }
    const errorMsg = getState().ui.routes.cart.errors.fewerThanOne.i18n.defaultText;
    const error = new BusinessLogicError(errorMsg);
    reject(error);
  })
    .then(() => {
      if (cb) {
        cb();
      }
    })
    .catch(error => {
      if (cb) {
        cb();
      }
      return Promise.reject(error);
    });
};

export const deleteCart = createAction(`${NS}RESET`);
reducer.on(deleteCart, state => {
  removeCartFromLocalStorage();
  return {
    ...state,
    itemsCount: 0,
    totalPrice: 0,
    items: [],
  };
});

export default reducer;
