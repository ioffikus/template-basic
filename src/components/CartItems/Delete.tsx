import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'antd';
import dispatch from 'src/store/dispatch';
import { deleteFromCart } from 'src/ducks/cart';
import Svg from 'src/components/Svg';

interface IOwnProps {
  mayDeleteItems: boolean;
  uid: string;
}

interface IConnectedState {}

type IProps = IOwnProps & IConnectedState;

interface IState {}

const mapStateToProps = (state: Store.IState) => ({});

const WrappedDelete = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Delete extends React.Component<IProps, IState> {
    handleDelete = (uid: string) => () => {
      dispatch(deleteFromCart(uid));
    };

    render() {
      const { mayDeleteItems, uid } = this.props;
      return (
        <div className="root">
          <style jsx>{`
            .root {
              text-align: right;
            }
            .root :global(.button-delete) {
              background: transparent;
              color: #b2bdcb;
            }
            .root :global(.icon-svg-trash) {
              height: 15px;
              width: 14px;
            }
          `}</style>
          {mayDeleteItems ? (
            <Button shape="circle" onClick={this.handleDelete(uid)} className="button-delete">
              <Svg name="trash" className="icon-svg-trash" />
            </Button>
          ) : (
            <span />
          )}
        </div>
      );
    }
  },
);

export default WrappedDelete;
