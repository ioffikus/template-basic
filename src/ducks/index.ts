// instead this, we have router in store via connected-react-router
// import { routerReducer as routing, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';
import auth from './auth';
import cart from './cart';
import cartPaidItems from './cartPaidItems';
import catalogQuery from './catalogQuery';
import catalog from './catalog';
import categories from './categories';
import categoriesMenu from './categoriesMenu';
import filter from './filter';
import locale from './locale';
import meta from './meta';
import page from './page';
import product from './product';
import progress from './progress';
import popularItems from './popularItems';
import recommendItems from './recommendItems';
import routes from './routes';
import ui from './ui';
import user from './user';
import userDelivery from './userDelivery';
import userOrders from './userOrders';

export default combineReducers<Store.IState>({
  ui,
  auth,
  user,
  categories,
  categoriesMenu,
  catalogQuery,
  catalog,
  product,
  progress,
  routes,
  meta,
  filter,
  cart,
  cartPaidItems,
  userDelivery,
  userOrders,
  popularItems,
  recommendItems,
  page,
  locale,
});
