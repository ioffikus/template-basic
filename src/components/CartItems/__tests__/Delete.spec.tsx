import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { Button } from 'antd';
import configureMockStore from 'redux-mock-store';
import { setupDispatch } from 'src/store/dispatch';
import { deleteFromCart } from 'src/ducks/cart';
import Delete from '../Delete';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
jest.mock('src/ducks/cart', () => {
  return {
    deleteFromCart: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});

describe(null, () => {
  it('render() with mayDeleteItems as true', () => {
    const mayDeleteItems = true;
    const uid = 'dummy';
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Delete mayDeleteItems={mayDeleteItems} uid={uid} />, {
      context: { store },
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('render() with mayDeleteItems as false', () => {
    const mayDeleteItems = false;
    const uid = 'dummy';
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Delete mayDeleteItems={mayDeleteItems} uid={uid} />, {
      context: { store },
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('Button.onClick() to Button', () => {
    const mayDeleteItems = true;
    const uid = 'dummy';
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Delete mayDeleteItems={mayDeleteItems} uid={uid} />, {
      context: { store },
    });
    const dummyWrapper = wrapper.dive();
    const buttonInstance = dummyWrapper
      .find(Button)
      .dive()
      .instance() as Button;
    buttonInstance.props.onClick(null);
    expect(deleteFromCart).toBeCalledWith(uid);
  });
});
