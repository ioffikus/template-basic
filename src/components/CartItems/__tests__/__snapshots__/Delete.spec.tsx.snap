// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() with mayDeleteItems as false 1`] = `
<div
  className="jsx-645318627 root"
>
  <JSXStyle
    css=".root.jsx-645318627{text-align:right;}.root.jsx-645318627 .button-delete{background:transparent;color:#b2bdcb;}.root.jsx-645318627 .icon-svg-trash{height:15px;width:14px;}"
    styleId="645318627"
  />
  <span
    className="jsx-645318627"
  />
</div>
`;

exports[`render() with mayDeleteItems as true 1`] = `
<div
  className="jsx-645318627 root"
>
  <JSXStyle
    css=".root.jsx-645318627{text-align:right;}.root.jsx-645318627 .button-delete{background:transparent;color:#b2bdcb;}.root.jsx-645318627 .icon-svg-trash{height:15px;width:14px;}"
    styleId="645318627"
  />
  <Button
    className="button-delete"
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    shape="circle"
  >
    <Svg
      className="icon-svg-trash"
      name="trash"
    />
  </Button>
</div>
`;
